serverless/email/webhook/main.go

```diff
@@ -0,0 +1,372 @@
+	}
+	return nil
+}
+
+// handler takes the api gateway request and sends a templated email
+func handler(ctx context.Context) func(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
+	logger := logging.Logger(ctx, "webhook.handler")
+	return func(request events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
+		var (
+			authenticated  bool
+			apiKey, authOK = request.Headers[keyHeaderName]
+			apiKeyHash     = sha256.Sum256([]byte(apiKey))
+		)
+
+		// no api key in request
+		if !authOK {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       http.StatusText(http.StatusUnauthorized),
+			}, nil
+		}
+
+		// check auth token against our comma seperated list of valid auth tokens
+		for _, token := range strings.Split(*authTokenSecretOutput.SecretString, ",") {
+			tokenHash := sha256.Sum256([]byte(token))
+			if subtle.ConstantTimeCompare(apiKeyHash[:], tokenHash[:]) == 1 {
+				authenticated = true
+			}
+		}
+
+		// api key in request does not match any configured
+		if !authenticated {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       http.StatusText(http.StatusUnauthorized),
+			}, nil
+		}
+
+		// validate the request signature
+		if err := validateSignature(request); err != nil {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusUnauthorized,
+				Body:       http.StatusText(http.StatusUnauthorized),
+			}, nil
+		}
+
+		// handler accepts from the request event the payload
+		// read the payload into our structure
+		payload := new(emailPayload)
+		err := json.Unmarshal([]byte(request.Body), payload)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to unmarshal request body")
+			// failed to unmarshal request appropriately
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusBadRequest,
+				Body:       http.StatusText(http.StatusBadRequest),
+			}, nil
+		}
+
+		// perform input payload validation
+		valid, err := govalidator.ValidateStruct(payload)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to validate the body structure")
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusInternalServerError,
+				Body:       http.StatusText(http.StatusInternalServerError),
+			}, nil
+		}
+
+		if !valid {
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusBadRequest,
+				Body:       http.StatusText(http.StatusBadRequest),
+			}, nil
+		}
+
+		// parse the namespace
+		namespace, err := uuid.Parse(*namespaceSecretOutput.SecretString)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to parse the namespace into a uuid")
+			return events.APIGatewayProxyResponse{
+				StatusCode: http.StatusInternalServerError,
+				Body:       http.StatusText(http.StatusInternalServerError),
+			}, nil
+		}
+
+		unsubscribeRef := uuid.NewSHA1(namespace, []byte(payload.Email)).String()
```

Comment:
Is SHA1 a requirement here or could we use at least SHA-256 here?

Reply:
This is based on the uuidv5 specification.  The unsubscribe reference is a uuid which is deterministically generated from the payload email from the webhook.

Reply:
👍🏻 gotcha makes sense for interop to keep it with sha-1. Technically highly unlikely we'd ever encounter a collision with this ever, so I think it's reasonable to keep this. Resolving this one as well